File: back-end/php/task_management/bll/base_service.php
<?php

class base_service {

    function init_user($user) {
        $new_user = array();
        $new_user['userId'] = $user['user_id'];
        $new_user['userName'] = $user['user_name'];
        $new_user['email'] = $user['email'];
        $new_user['password'] = $user['password'];
        $new_user['profileImageName'] = $user['profile_image_name'];
        $new_user['departmentId'] = $user['department_id'];
        $new_user['teamLeaderId'] = $user['team_leader_id'];
        $new_user['managerId'] = $user['manager_id'];
        $new_user['isActive'] = $user['is_active'];
        if (array_key_exists('department_name', $user)) {
            $new_user['department'] = array();
            $new_user['department']['departmentName'] = $user['department_name'];
        }
        if (array_key_exists('team_leader_name', $user)) {
            $new_user['teamLeader'] = array();
            $new_user['teamLeader']['userName'] = $user['team_leader_name'];
        }
        return $new_user;
    }

    function init_project($project) {
        $new_project = array();
        $new_project['projectId'] = $project['project_id'];
        $new_project['projectId'] = $project['project_id'];
        $new_project['projectId'] = $project['project_id'];
        $new_project['projectName'] = $project['project_name'];
        $new_project['managerId'] = $project['manager_id'];
        $new_project['customerId'] = $project['customer_id'];
        $new_project['teamLeaderId'] = $project['team_leader_id'];
        $new_project['totalHours'] = $project['total_hours'];
        $new_project['startDate'] = $project['start_date'];
        $new_project['endDate'] = $project['end_date'];
        $new_project['isComplete'] = $project['is_complete'];
        if (array_key_exists('customer_name', $project)) {
            $new_project['customer'] = array();
            $new_project['customer']['customerName'] = $project['customer_name'];
        }
        if (array_key_exists('user_name', $project) && array_key_exists('email', $project)) {
            $new_project['teamLeader'] = array();
            $new_project['teamLeader']['userName'] = $project['user_name'];
            $new_project['teamLeader']['email'] = $project['email'];
        }
        return $new_project;
    }

    function init_customer($customer) {
        $new_customer = array();
        $new_customer['customerId'] = $customer['customer_id'];
        $new_customer['customerName'] = $customer['customer_name'];
        return $new_customer;
    }

    function init_department($department) {
        $new_department = array();
        $new_department['departmentId'] = $department['department_id'];
        $new_department['departmentName'] = $department['department_name'];
        return $new_department;
    }

    function init_worker_hours($worker_hours) {
        $new_worker_hours = array();
        $new_worker_hours['workerHoursId'] = $worker_hours['worker_hours_id'];
        $new_worker_hours['projectId'] = $worker_hours['project_id'];
        $new_worker_hours['workerId'] = $worker_hours['worker_id'];
        $new_worker_hours['numHours'] = $worker_hours['num_hours'];
        $new_worker_hours['isComplete'] = $worker_hours['is_complete'];
        $new_worker_hours['project'] = array();
        $new_worker_hours['project']['projectName'] = $worker_hours['project_name'];
        $new_worker_hours['worker'] = array();
        $new_worker_hours['worker']['userName'] = $worker_hours['user_name'];
        $new_worker_hours['worker']['email'] = $worker_hours['email'];
        $new_worker_hours['worker']['department'] = array();
        $new_worker_hours['worker']['department']['departmentName'] = $worker_hours['department_name'];
        return $new_worker_hours;
    }

    function init_department_hours($department_hours) {
        $new_department_hours = array();
        $new_department_hours['departmentHoursId'] = $department_hours['department_hours_id'];
        $new_department_hours['projectId'] = $department_hours['project_id'];
        $new_department_hours['departmentId'] = $department_hours['department_id'];
        $new_department_hours['numHours'] = $department_hours['num_hours'];
        $new_department_hours['department'] = array();
        $new_department_hours['department']['departmentName'] = $department_hours['department_name'];
        return $new_department_hours;
    }

    function init_permission($permission) {
        $new_permission = array();
        $new_permission['permissionId'] = $permission['permission_id'];
        $new_permission['workerId'] = $permission['worker_id'];
        $new_permission['projectId'] = $permission['project_id'];
        $new_permission['isActive'] = $permission['is_active'];
        $new_permission['worker'] = array();
        $new_permission['worker']['userName'] = $permission['user_name'];
        $new_permission['project'] = array();
        $new_permission['project']['projectName'] = $permission['project_name'];
        return $new_permission;
    }

    function init_presence_hours($presence_hours) {
        $new_presence_hours = array();
        $new_presence_hours['presenceHoursId'] = $presence_hours['presence_hours_id'];
        $new_presence_hours['workerId'] = $presence_hours['worker_id'];
        $new_presence_hours['projectId'] = $presence_hours['project_id'];
        $new_presence_hours['startHour'] = $presence_hours['start_hour'];
        $new_presence_hours['endHour'] = $presence_hours['end_hour'];
        $new_presence_hours['worker'] = array();
        $new_presence_hours['worker'] ['userName'] = $presence_hours['user_name'];
        return $new_presence_hours;
    }

    function format_date($date, $format = 'Y-m-d') {
        $format_date = date($format, strtotime($date));
        return "'$format_date'";
    }

    function get_value_or_null($array, $key) {
        return array_key_exists($key, $array) && isset($array[$key]) ? "'{$array[$key]}'" : 'null';
    }

}

File: back-end/c#/TaskManagement/BLL/BaseService.cs
using BOL;
using BOL.Help;
using System;
using System.Text;
using System.Net.Mail;
using System.Net;
using MySql.Data.MySqlClient;
using System.Collections.Generic;
using System.Linq;
using System.Timers;
using System.Configuration;
using System.Threading;
using System.Threading.Tasks;

namespace BLL
{
    public static class BaseService
    {
        public static string GetStringValueOrNull(string value)
        {
            return value != null ? $"'{value}'" : "null";
        }

        public static string GetIntValueOrNull(int? value)
        {
            return value != null ? $"{value}" : "null";
        }
        public static string FormatDate(DateTime date,string format="yyyy-MM-dd")
        {
            return $"'{date.ToString(format)}'";
        }

        public static bool SendEmail(Email email)
        {
            try
            {
                SmtpClient smtpClient = new SmtpClient("smtp.gmail.com", 587);
                smtpClient.EnableSsl = true;
                smtpClient.Timeout = 1000000;
                smtpClient.UseDefaultCredentials = false;

                NetworkCredential basicCredential = new NetworkCredential(email.UserName,email.Password);
                smtpClient.Credentials = basicCredential;

                MailMessage mailMessage = new MailMessage();
                mailMessage.From = new MailAddress(email.UserName);
                email.ToAddress.ForEach(toAddress => mailMessage.To.Add(toAddress));
                mailMessage.Subject = email.Subject;
                mailMessage.Body = email.Body;
                mailMessage.BodyEncoding = Encoding.UTF8;
                smtpClient.Send(mailMessage);
                return true;
            }
            catch
            {
                return false;
            }
        }
        public static async Task DeadLineEmail()
        {
            // Set up a timer that triggers every day.
            System.Timers.Timer timer = new System.Timers.Timer();
            //timer.Interval = 60000 * 60 * 24; // one day
            timer.Interval = 60000;

            timer.Elapsed += new ElapsedEventHandler(OnStart);

            string time = ConfigurationManager.AppSettings["deadLineEmailHour"];
            int hour = int.Parse(time.Split(':')[0]);
            int minute = int.Parse(time.Split(':')[1]);

            DateTime currentDateTime = DateTime.Now;
            DateTime dateTimeToStart = currentDateTime.Date + new TimeSpan(hour, minute, 0);
            if (dateTimeToStart < currentDateTime)
                dateTimeToStart= dateTimeToStart.AddDays(1);
            TimeSpan timeout = dateTimeToStart - currentDateTime;
            Thread.Sleep(timeout);
            timer.Start();
        }

        public static void OnStart(object sender, ElapsedEventArgs args)
        {
            sendDeadLineEmail();
        }

        public static void sendDeadLineEmail()
        {
            List<Project> projects = ProjectService.GetAllProjects(DateTime.Today);
            projects.ForEach(project =>
            {
                Email emailToTeamLeader = new Email() { Subject = "Deadline is Here!!", Body = $"Your project {project.ProjectName} is finish tomorrow!\n please pay attention to verify your workers finished their task." };
                Email emailToWorkers = new Email() { Subject = "Deadline is Here!!", Body = $"Your project {project.ProjectName} is finish tomorrow!\n please pay attention to finish your task." };

                List<WorkerHours> workerHoursList = WorkerHoursService.GetUncompletedWorkerHours(project.ProjectId);
                List<string> emailAddresses = workerHoursList.Select(workerHours => workerHours.Worker.Email).ToList();
                emailToWorkers.ToAddress.AddRange(emailAddresses);
                if (workerHoursList.Count > 0)
                    emailToTeamLeader.ToAddress.Add(UserService.GetUserById(project.TeamLeaderId).Email);
                SendEmail(emailToWorkers);
                SendEmail(emailToTeamLeader);
            });
        }
       
        public static Customer InitCustomer(MySqlDataReader reader)
        {
            try
            {
                Customer customer = new Customer
                {
                    CustomerId = reader.GetInt32(0),
                    CustomerName = reader.GetString(1)
                };
                return customer;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static Department InitDepartment(MySqlDataReader reader)
        {
            try
            {
                Department department = new Department
                {
                    DepartmentId = reader.GetInt32(0),
                    DepartmentName = reader.GetString(1),
                };
                return department;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public static Permission InitPermission(MySqlDataReader reader)
        {
            try
            {
                Permission Permission = new Permission
                {
                    PermissionId = reader.GetInt32(0),
                    WorkerId = reader.GetInt32(1),
                    ProjectId = reader.GetInt32(2),
                    IsActive = reader.GetBoolean(3),
                    Worker = new User() { UserName = reader.GetString(4) },
                    Project = new Project() { ProjectName = reader.GetString(5) }
                };
                return Permission;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }


        public static PresenceHours initPresenceHours(MySqlDataReader reader)
        {
            try
            {
                PresenceHours presenceHours = new PresenceHours
                {
                    PresenceHoursId = reader.GetInt32(0),
                    WorkerId = reader.GetInt32(1),
                    ProjectId = reader.GetInt32(2),
                    StartHour = reader.GetDateTime(3),
                    EndHour = reader.GetDateTime(4),
                    Worker = new User() { UserName = reader.GetString(5) }
                };
                return presenceHours;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public static Project InitProject(MySqlDataReader reader)
        {
            Project project = new Project
            {
                ProjectId = reader.GetInt32(0),
                ProjectName = reader.GetString(1),
                ManagerId = reader.GetInt32(2),
                CustomerId = reader.GetInt32(3),
                TeamLeaderId = reader.GetInt32(4),
                TotalHours = reader.GetInt32(5),
                StartDate = reader.GetDateTime(6),
                EndDate = reader.GetDateTime(7),
                IsComplete=reader.GetBoolean(8)
            };
            if (reader.FieldCount == 12)
            {
                project.Customer = new Customer { CustomerName = reader.GetString(9) };
                project.TeamLeader = new User { UserName = reader.GetString(10), Email = reader.GetString(11) };
            }
            return project;
        }

        public static User InitUser(MySqlDataReader reader)
        {
            try
            {
                User user = new User
                {
                    UserId = reader.GetInt32(0),
                    UserName = reader.GetString(1),
                    Email = reader.GetString(2),
                    Password = reader.GetString(3),
                    ProfileImageName = reader.IsDBNull(4) ? default(string) :
                                  reader.GetString(4),
                    IsManager = reader.GetBoolean(5),
                    DepartmentId = reader.IsDBNull(6) ? default(int?) :
                                      reader.GetInt32(6),
                    TeamLeaderId = reader.IsDBNull(7) ? default(int?) :
                                      reader.GetInt32(7),
                    ManagerId = reader.IsDBNull(8) ? default(int?) :
                                      reader.GetInt32(8),
                    IsActive = reader.GetBoolean(9),
                };
                if (reader.FieldCount == 12 && reader.IsDBNull(10) == false)
                {
                    user.Department = new Department { DepartmentName = reader.GetString(10) };
                }
                if (reader.FieldCount == 12 && reader.IsDBNull(11) == false)
                {
                    user.TeamLeader = new User { UserName = reader.GetString(11) };
                }
                return user;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public static WorkerHours InitWorkerHours(MySqlDataReader reader)
        {
            try
            {
                WorkerHours workerHours = new WorkerHours
                {
                    WorkerHoursId = reader.GetInt32(0),
                    ProjectId = reader.GetInt32(1),
                    WorkerId = reader.GetInt32(2),
                    NumHours = reader.GetInt32(3),
                    IsComplete=reader.GetBoolean(4),
                    IsActive = reader.GetBoolean(5),
                    Project = new Project { ProjectName = reader.GetString(6) },
                    Worker = new User { UserName = reader.GetString(7),Email=reader.GetString(8), Department = new Department { DepartmentName = reader.GetString(9) } }
                };

                return workerHours;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static DepartmentHours InitDepartmentHours(MySqlDataReader reader)
        {
            DepartmentHours departmentHours = new DepartmentHours
            {
                DepartmentHoursId = reader.GetInt32(0),
                ProjectId = reader.GetInt32(1),
                DepartmentId = reader.GetInt32(2),
                NumHours = reader.GetInt32(3),
                Department = new Department { DepartmentName = reader.GetString(4) }
            };
            return departmentHours;
        }
    }
}
